unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,128 @@
+import json
+import multiprocessing as mp
+import typing as t
+from dataclasses import dataclass
+
+from unstructured.ingest.error import DestinationConnectionError, WriteError
+from unstructured.ingest.interfaces import (
+    BaseConnectorConfig,
+    BaseDestinationConnector,
+    BaseIngestDoc,
+    BaseSessionHandle,
+    ConfigSessionHandleMixin,
+    WriteConfig,
+    WriteConfigSessionHandleMixin,
+)
+from unstructured.ingest.logger import logger
+from unstructured.staging.base import flatten_dict
+from unstructured.utils import requires_dependencies
+
+
+@dataclass
+class PineconeSessionHandle(BaseSessionHandle):
+    service: "pinecone.Index"  # noqa: F821
+
+
+@dataclass
+class PineconeWriteConfig(WriteConfigSessionHandleMixin, ConfigSessionHandleMixin, WriteConfig):
+    api_key: str
+    index_name: str
+    environment: str
+    batch_size: str
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def create_pinecone_object(self, api_key, index_name, environment):
+        import pinecone
+
+        pinecone.init(api_key=api_key, environment=environment)
+        index = pinecone.Index(index_name)
+        logger.debug(f"Connected to index: {pinecone.describe_index(index_name)}")
+        return index
+
+    def create_session_handle(self) -> PineconeSessionHandle:
+        service = self.create_pinecone_object(self.api_key, self.index_name, self.environment)
+        return PineconeSessionHandle(service=service)
+
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        index = self.session_handle.service
+        try:
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.ApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
+
+
+@dataclass
+class SimplePineconeConfig(BaseConnectorConfig):
+    api_key: str
+    index_name: str
+    environment: str
+
+
+@dataclass
+class PineconeDestinationConnector(BaseDestinationConnector):
+    write_config: PineconeWriteConfig
+    connector_config: SimplePineconeConfig
+
+    def initialize(self):
+        pass
+
+    def check_connection(self):
+        pass
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(
+            f"Inserting / updating {len(dict_list)} documents to destination "
+            f"index at {self.connector_config.index_name}",
+        )
+
+        # this is advised to be 100 at maximum in pinecone docs, however when we
+        # chunk content, we hit to the object size limits, so we decrease the batch
+        # size even more here
+        pinecone_batch_size = self.write_config.batch_size
+
+        num_processes = 1
```

Comment:
Was this set here for development? Or is there a problem with multi-processing?

Reply:
+1 👆

Reply:
Was for development indeed. Now, I've exposed an additional `num_processes` for the writer, and have set the test to do parallel uploads: 
[expose writer num_processes, apply parallelization in ingest test](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/1ff1fd665ce0499be71b91149927f39288d7f17d)

Here are some logs in local (processing 3 files here) for reference. Batch with the smaller ("remainder") amount being uploaded earlier means that the batches are not being processed in order; pointing to successful parallelization:
```
2023-11-09 13:06:32,485 MainProcess INFO     appending 558 json elements from content in .../unstructured/test_unstructured_ingest/structured-output/s3-pinecone-dest/book-war-and-peace-1225p.txt.json
2023-11-09 13:06:32,503 MainProcess INFO     appending 174 json elements from content in .../unstructured/test_unstructured_ingest/structured-output/s3-pinecone-dest/example-10k-230p.html.json
2023-11-09 13:06:32,503 MainProcess INFO     appending 2 json elements from content in .../unstructured/test_unstructured_ingest/structured-output/s3-pinecone-dest/DA-1p.pdf.json
2023-11-09 13:06:32,503 MainProcess INFO     Upserting 734 elements to destination index at ingest-test
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 14}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
results: {'upserted_count': 80}
```